pythonforandroid/bootstraps/common/build/jni/application/src/start.c

```diff
@@ -243,9 +275,69 @@ int main(int argc, char *argv[]) {
 #if PY_MAJOR_VERSION < 3
   PyRun_SimpleString("import site; print site.getsitepackages()\n");
 #endif
-
   LOGP("AND: Ran string");
 
+#if PY_MAJOR_VERSION >= 3
+  /* Workaround crash in ctypes.find_library() which is caused by its
+   * internal use of /bin/sh, and looking in the wrong folders on Android.
+   *
+   * The code needs 'env_argument' as baked path, so we need to insert this:
+   */
+  LOGP("ctypes workaround insert");
+  char code_with_placeholder[] = (""
+    "try:\n"
+    "    import ctypes.util\n"
+    "    orig_func = ctypes.util.find_library\n"
+    "    def android_find_library_hack(*args):\n"
+    "        import os\n"
+    "        name = args[0]\n"
+    "        \n"
+    "        # Truncate ending for easier comparison:\n"
+    "        if name.find('.so.') >= 0:\n"
+    "            name = name.partition('.so.')[0]\n"
+    "        if name.endswith('.so'):\n"
+    "            name = name[:-len('.so')]\n"
+    "        if not name.endswith('.'):\n"
+    "            name += '.'\n"
+    "        \n"
+    "        # Helper function to check lib name:\n"
+    "        def check_name(lib_name, search_name):\n"
+    "            if filename.endswith('.so') and (\n"
+    "                    filename.startswith('lib' + search_name) or\n"
+    "                    filename.startswith(search_name)):\n"
+    "                return True\n"
+    "            if search_name.endswith('-2.0.'):\n" // PySDL2 hack
+    "                search_name = search_name[:-len('-2.0.')] + '.'\n"
+    "                return check_name(lib_name, search_name)\n"
+    "            return False\n"
+    "        \n"
+    "        # Check the user app lib dir and system dir:\n"
+    "        app_root = os.path.normpath(os.path.abspath(os.path.join(\n"
+    "            '%s', '..', '..', 'lib')))\n"
+    "        app_root_arm = os.path.join(app_root,\n"
+    "                                    'arm')  # fixme: other archs?\n"
+    "        sys_dir = '/system/lib'\n"
+    "        for lib_search in [app_root, app_root_arm, sys_dir]:\n"
+    "            if not os.path.exists(lib_search):\n"
+    "                continue\n"
+    "            for filename in os.listdir(lib_search):\n"
+    "                if check_name(filename, name):\n"
+    "                    return os.path.join(lib_search, filename)\n"
+    "        return orig_func(*args)\n"
+    "    ctypes.util.find_library = android_find_library_hack\n"
+    "except ImportError:\n"
+    "    pass\n");
```

Comment:
Maybe could log something here, like "Could not import ctypes, build your app with libffi if you want ctypes to be present". I think that could help people.

Reply:
I put that in now!

I'm just curious though, is `libffi` shipped per default and would need to be opted out of with some option, or does it actually need to be put into `--requirements` for ctypes to be present at all? Because an opt-out intuitively would seem to make more sense to me, in which case people would already be aware and the warning wouldn't be that important, would it?